refactor loader into a type and hide it behind Datapath interface

ianvernon opened the pull request:
Read per-commit.

The goal here is to abstract out the datapath interactions away from the endpoint / daemon, so that we can do more unit / mock-based testing around things which use the datapath. There are no functional changes intended as part of this PR. 

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8769)
<!-- Reviewable:end -->

ianvernon commented on pkg/datapath/loader/wrapper/interface.go:
```diff
@@ -0,0 +1,41 @@
+	"github.com/cilium/cilium/pkg/datapath/loader/metrics"
+)
+
+type LoaderWrapper struct{}
```
Will add documentation later.

nebril commented on pkg/endpoint/endpoint.go:
```diff
@@ -297,6 +301,10 @@ type Endpoint struct {
 
+	// DatapathLoaderImpl is responsible for the compilation / reloading of the
+	// datapath for this endpoint.
+	DatapathLoaderImpl DatapathLoader `json:"-"`
```
Can this field be called `DatapathLoader` or just `Loader`?

ianvernon replied:
Renamed to `Loader`.

nebril commented on pkg/endpoint/metrics.go:
```diff
@@ -18,7 +18,7 @@ import (
 
 	"github.com/cilium/cilium/api/v1/models"
-	"github.com/cilium/cilium/pkg/datapath/loader"
+	metrics2 "github.com/cilium/cilium/pkg/datapath/loader/metrics"
```
I would rather have `loaderMetrics` than `metrics2 here

ianvernon replied:
Done!

nebril reviewed (commented):
Looks good overall aside from naming issues.

joestringer commented on pkg/datapath/loader/config/config.go:
```diff
@@ -0,0 +1,16 @@
+
+// EndpointConfig provides access to EndpointConfig information that is necessary to
+// compile and load the datapath.
+type EndpointConfig interface {
```
What's different conceptually between this `EndpointConfig` and  the nested `datapath.EndpointConfiguration`? They have very similar names.

Can (should?) we just extend the datapath.EndpointConfiguration type to have these extra bits since this new package is introducing the dependency on the datapath package anyway?

ianvernon replied:
I factored out the `endpoint` interface out of `pkg/loader` into its own `Endpoint` interface within `pkg/datapath`. I'll look into seeing if I can collapse the two interfaces together, that's a good idea.

joestringer commented on pkg/datapath/loader/wrapper/interface.go:
```diff
@@ -0,0 +1,41 @@
+	"github.com/cilium/cilium/pkg/datapath/loader/metrics"
+)
+
+type LoaderWrapper struct{}
```
Is there a reason to introduce this wrapper type rather than just convert the existing loader into this form with a singleton-style object like this that has methods?

ianvernon replied:
I substantially changed this in the most recent upload. Let me know what you think. A `Datapath` now has a `Loader`. We get this `Loader` in the endpoint by accessing the `Datapath` of the endpoint's `owner`.

joestringer reviewed (commented):
This seems like potentially a good direction, but I don't think the implementation is going far enough. Seems like we're introducing extra unnecessary types when we should just go ahead and change the core objects/implementations themselves.

Commit:
datapath: factor out configuration-writing functions into interface

A `Datapath` is composed of a `ConfigWriter`, so no functional changes are
present in this commit. Future commits depend on this factoring out so that
there are not import cycles.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
loader: change parameters to take `ConfigWriter` instead of `Datapath`

This ensures that in latter commits, there are not import cycles.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
datapath/linux: factor out configuration writing into separate package

This ensures that in latter commits, there are not import cycles.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
loader: create `Loader` type

Create functions that have a `*Loader` as a receiver. Maintain package-level
functionality to avoid breaking users of `pkg/loader` throughout cilium for now
by having a package-level loader variable, and have the package-level functions
work on this package-level structure.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
loader: refactor loader metrics into separate package

This ensures that there are not import cycles in latter commits.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
factor out endpoint interface from loader to pkg/datapath

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
datapath: add Loader interface

This is currently only implemented by `pkg/loader:Loader`.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
datapath: add `Loader` function to Datapath interface

Add this function to both implementing types (`pkg/linux/datapath:linuxDatapath`
, and `pkg/datapath/fake:fakeDatapath`). The linux datapath now has a
`*loader.Loader` field

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
cilium: replace use of package-level loader functions

Access via the Datapath() field of the Daemon / Owner now.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
loader: remove globalLoader and associated package-level functions

These aren't used any more, so remove them accordingly.

Signed-off by: Ian Vernon <ian@cilium.io>

raybejjani commented on pkg/datapath/linux/datapath.go:
```diff
@@ -72,3 +77,19 @@ func (l *linuxDatapath) InstallProxyRules(proxyPort uint16, ingress bool, name s
 	return iptables.RemoveProxyRules(proxyPort, ingress, name)
 }
+
+func (l *linuxDatapath) WriteTemplateConfig(w io.Writer, e datapath.EndpointConfiguration) error {
```
I think I missed something but we seem to use the configWriter directly in other places, where does linuxDatapath need to implement ConfigWriter?
I'm guessing there is some spot where we still want this to meet the ConfigWriter interface, but couldn't we just use l.configWriter and pass that out? (or, always use the linuxDatapath itself but I think that's the exact loop you're trying to avoid).

ianvernon replied:
I guess `linuxDatapath` is a `ConfigWriter` since the `Datapath` interface is composed with the `ConfigWriter` interface. But, the crux of your question seems to be why does a `Datapath` have to be composed with a `ConfigWriter` - is that correct? 

The endpoint package uses the `Datapath` of its `Owner` to do these data path configuration updates, and I didn't want to change that up in this PR.

Let me know if I am on the right track - we can discuss offline as well, as I want to make sure I am grasping your question correctly.

joestringer replied:
My question looking at this is why do we need to write all this boilerplate? Why not just publicly compose the inner ConfigWriter?

ianvernon replied:
can you provide an example of what you mean by "publicly compose"? e.g., `Datapath().ConfigWriter()`?

ianvernon replied:
Discussed offline with @joestringer  - `linuxDatapath` is now composed with a `ConfigWriter` . @raybejjani is this what you had in mind?

raybejjani reviewed (approved):
LGTM

Commit:
loader: make linuxDatapath composed of a ConfigWriter

This removes unnecessary boilerplate around the implementation of the `ConfigWriter` interfaces.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -59,7 +58,7 @@ func (e *Endpoint) PolicyMapPathLocked() string {
 // CallsMapPathLocked returns the path to cilium tail calls map of an endpoint.
 func (e *Endpoint) CallsMapPathLocked() string {
-	return bpf.LocalMapPath(loader.CallsMapName, e.ID)
+	return e.owner.Datapath().Loader().CallsMapPath(e.ID)
```
Looking at these changes, I do wonder why the endpoint package needs to know the chain of objects from the owner down to figuring out the CallsMapPath. Though that said for this particular case I'd imagine we can get rid of these map paths helpers from endpoint in a next step of iteration on the endpoint datapath abstractions.

ianvernon replied:
Agreed - I think these refactors will review some odd dependency chains that we have which will motivate other refactors as well.
